File: 백준/Gold/1043. 거짓말/거짓말.cc
#include <bits/stdc++.h>

using namespace std;

int n, m;
int k;
int know_people[51] = { 0, };
int result = 0;
int visited[51] = { 0, };
int meet[51][51] = { 0, };
vector<int> party[51];
queue<int> q;


void bfs()
{
	while (!q.empty())
	{
		int cur = q.front();

		q.pop();

		visited[cur] = 1;
		know_people[cur] = 1;

		for (int i = 1; i <= n; i++)
		{
			if (meet[cur][i] == 1 && !visited[i])
			{
				know_people[i] = 1;
				q.push(i);
			}
		}
	}
}

void solve()
{
	bfs();

	for (int i = 0; i < m; i++)
	{
		int can_gura = 1;
		for (int j = 0; j < party[i].size(); j++)
		{
			if (know_people[party[i][j]])
				can_gura = 0;
		}

		if (can_gura)
		{
			result++;
		}
	}
}


int main()
{
	cin >> n >> m >> k;

	for (int i = 0; i < k; i++)
	{
		int t;
		cin >> t;
		know_people[t] = 1;
		q.push(t);
	}

	for (int i = 0; i < m; i++)
	{
		// 몇명이 파티에 참석하는지?
		int num;
		cin >> num;
		for (int j = 0; j < num; j++)
		{
			int temp;
			cin >> temp;
			party[i].push_back(temp);
		}

		// 누가 누굴 만나는지 기록
		for (int j = 0; j < num; j++)
		{
			for (int k = 0; k < num; k++)
			{
				if (j != k)
					meet[party[i][j]][party[i][k]] = 1;
			}
		}
	}

	solve();

	cout << result;

	return 0;
}

File: 백준/Gold/1043. 거짓말/거짓말.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static int n, m, t, result;
    static int[] know, visited;
    static int[][] Map;
    static ArrayList<Integer>[] party;

    public static void main(String[] args) throws IOException {

        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());

        t = Integer.parseInt(st.nextToken());

        know = new int[t];
        visited = new int[51];
        Map = new int[51][51];
        party = new ArrayList[m];
        for (int i = 0 ; i < m ; i ++) {
            party[i] = new ArrayList<>();
        }

        for (int i = 0; i < t ; i ++) {
            know[i] = Integer.parseInt(st.nextToken());
        }

        for (int i = 0 ; i < m ; i ++) {
            st = new StringTokenizer(br.readLine());

            int cnt = Integer.parseInt(st.nextToken());
            int start = 0;
            for (int j = 0 ; j < cnt ; j ++) {
                int val = Integer.parseInt(st.nextToken());
                party[i].add(val);
                if (j == 0) {
                    start = val;
                } else {
                    Map[start][val] = 1;
                    Map[val][start] = 1;
                }
            }
        }

        bfs();

        for (int i = 0 ; i < m; i ++) {
            boolean cnt = true;
            for (int j = 0; j < party[i].size(); j++) {
                if (visited[party[i].get(j)] == 1) {
                    cnt = false;
                    break;
                }
            }

            if (cnt)
                result++;
        }

        System.out.println(result);
    }

    static void bfs() {
        Queue<Integer> q = new LinkedList<>();
        for (int i = 0 ; i < t ; i++) {
            q.add(know[i]);
            visited[know[i]] = 1;
        }

        while(!q.isEmpty()) {
            int cur = q.poll();

            for (int i = 1; i <= 50 ; i++) {
                if (Map[cur][i] == 1 && visited[i] == 0) {
                    q.add(i);
                    visited[i] = 1;
                }
            }
        }
    }
}
